src/tezWrapper.ml

```diff
@@ -287,14 +291,45 @@ let[@inline] set_delegate (state: tez_wrapper_state) (kho: Ligo.key_hash option)
   let _ = ensure_no_tez_given () in
   (* 2. Create a vault, if it does not exist already, and update the map. *)
   let op_opt, state, vault_address = find_vault_address state !Ligo.Tezos.sender in
-  (* 3. Instruct the vault to set its own delegate *)
-  let op = match (LigoOp.Tezos.get_entrypoint_opt "%vault_set_delegate" vault_address : Ligo.key_hash option Ligo.contract option) with
-    | Some c -> LigoOp.Tezos.opt_key_hash_transaction kho (Ligo.tez_from_literal "0mutez") c
-    | None -> (Ligo.failwith error_GetEntrypointOptFailureVaultSetDelegate : LigoOp.operation) in
+  (* 3. Instruct the vault to set its own delegate (via indirect internal call). *)
+  let op = match (LigoOp.Tezos.get_entrypoint_opt "%call_vault_set_delegate" !Ligo.Tezos.self_address : (Ligo.address * Ligo.key_hash option) Ligo.contract option) with
+    | Some c -> LigoOp.Tezos.address_opt_key_hash_transaction (vault_address, kho) (Ligo.tez_from_literal "0mutez") c
+    | None -> (Ligo.failwith error_GetEntrypointOptFailureCallVaultSetDelegate : LigoOp.operation) in
   match op_opt with
   | None -> ([op], state)
   | Some origination -> ([origination; op], state)
 
+(*****************************************************************************)
+(**                      {1 INTERNAL ENTRYPOINTS}                            *)
+(*****************************************************************************)
+
+let[@inline] call_vault_receive_tez (state: tez_wrapper_state) (vault_address, amnt : Ligo.address * Ligo.tez) : LigoOp.operation list * tez_wrapper_state =
+  if Ligo.Tezos.sender <> Ligo.Tezos.self_address then
+    (Ligo.failwith error_UnauthorisedCaller : LigoOp.operation list * tez_wrapper_state)
+  else
+    let op = match (LigoOp.Tezos.get_entrypoint_opt "%vault_receive_tez" vault_address : unit Ligo.contract option) with
+      | Some c -> LigoOp.Tezos.unit_transaction () amnt c
+      | None -> (Ligo.failwith error_GetEntrypointOptFailureVaultReceiveTez : LigoOp.operation) in
+    ([op], state)
+
+let[@inline] call_vault_send_tez_to_contract (state: tez_wrapper_state) (vault_address, amnt, recipient : Ligo.address * Ligo.tez * Ligo.address) : LigoOp.operation list * tez_wrapper_state =
+  if Ligo.Tezos.sender <> Ligo.Tezos.self_address then
+    (Ligo.failwith error_UnauthorisedCaller : LigoOp.operation list * tez_wrapper_state)
+  else
+    let op = match (LigoOp.Tezos.get_entrypoint_opt "%vault_send_tez_to_contract" vault_address : (Ligo.tez * Ligo.address) Ligo.contract option) with
+      | Some c -> LigoOp.Tezos.tez_address_transaction (amnt, recipient) (Ligo.tez_from_literal "0mutez") c
+      | None -> (Ligo.failwith error_GetEntrypointOptFailureVaultSendTezToContract : LigoOp.operation) in
+    ([op], state)
+
+let[@inline] call_vault_set_delegate (state: tez_wrapper_state) (vault_address, kho : Ligo.address * Ligo.key_hash option) : LigoOp.operation list * tez_wrapper_state =
+  if Ligo.Tezos.sender <> Ligo.Tezos.self_address then
```

Comment:
:+1:

Reply:
Ah, you have sharper eyes than me today. Nice catch and sorry I didn't notice that in the review :grimacing:

Reply:
Ha, no worries! Strictly-speaking this is not a bug; the final michelson will be correct. It's just a weakness of our compatibility layer, what can you do! :shrug: